### PR TITLE
Allow old-style exception types for backported Future.

### PR DESCRIPTION
**Pull request by rhansen**
`third_party/2/concurrent/futures` is for the Python 2 backport of the `concurrent.futures` package from Python 3.  In Python 3, all exceptions are derived from `BaseException`, but Python 2 has exceptions that are old-style objects.  Update the exception type for `concurrent.futures.Future.set_exception()` and `.exception()` to include `types.InstanceType`, which is the type of all old-style objects.

### CODE REVIEW

File: third_party/2/concurrent/futures/__init__.pyi [hunk 1]
```diff
@@ -1,5 +1,9 @@
+import types
 from typing import TypeVar, Generic, Any, Iterable, Iterator, Callable, Tuple, Union
 
+// Old-style exceptions (which were removed in Python 3.x) are instances of
```
Comment (ambv) [inline third_party/2/concurrent/futures/__init__.pyi]: Linter is right, this is a syntax error.

File: third_party/2/concurrent/futures/__init__.pyi [hunk 2]
```diff
@@ -1,5 +1,9 @@
...
+// types.InstanceType.  See https://github.com/agronholm/pythonfutures/pull/58
+_ExceptionType = Union[BaseException, types.InstanceType]
```
Comment (ambv) [inline third_party/2/concurrent/futures/__init__.pyi]: This basically means "any old-style class or BaseException" so it provides a very broad spectrum. I'm tempted to just advise putting "Any" instead and be done with it.

File: third_party/2/concurrent/futures/__init__.pyi [hunk 3]
```diff
@@ -8,12 +12,12 @@ class Future(Generic[_T]):
     def running(self) -> bool: ...
     def done(self) -> bool: ...
     def result(self, timeout: float = ...) -> _T: ...
-    def exception(self, timeout: float = ...) -> BaseException: ...
+    def exception(self, timeout: float = ...) -> _ExceptionType: ...
```
Comment (ambv) [inline third_party/2/concurrent/futures/__init__.pyi]: Don't return Unions, just "Any".

See, the way Mypy treats those cases is a little unintuitive:
- a function returning `Union[A, B]` can return either type A or type B; so
- code using that return value is only safe if it uses the value in places safe for both types; for example
- using the return value's methods and attributes is only okay if they are present in all types within the union (everything else will raise an error).

In our case the union is with any old-style class. So it means only methods and attributes present on both old-style classes and BaseExceptions are acceptable.

File: third_party/2/concurrent/futures/__init__.pyi (hunk 1 above)
Comment (rhansen) [inline third_party/2/concurrent/futures/__init__.pyi]: Well that's just embarrassing.  Too much C++ lately.  Fixed.

File: third_party/2/concurrent/futures/__init__.pyi (hunk 3 above)
Comment (rhansen) [inline third_party/2/concurrent/futures/__init__.pyi]: I don't quite follow your arguments against the `Union`.

The `exception()` method can return any arbitrary old-style object, so it's correct to say that callers can only safely use attributes that are present on both old-style objects and `BaseException`.  The `Union` helps enforce this restriction.

File: third_party/2/concurrent/futures/__init__.pyi (hunk 2 above)
Comment (rhansen) [inline third_party/2/concurrent/futures/__init__.pyi]: It is broad, but it's narrower than `Any`.

**Comment by rhansen** [comment]
friendly ping

**Comment by gvanrossum** [comment]
I've tried to write a simple demo program that raises an exception that doesn't derive from `BaseException`, and it doesn't even work -- the thread just crashes. Now, I suppose you could set the exception using `set_exception()` but I think that's a very rare use case, given for how long Python 2 has strongly recommended deriving exceptions from `BaseException`.

Here's my code:
```
from __future__ import print_function
from concurrent.futures import ThreadPoolExecutor
import time
import os

MYPY = False
if not MYPY:
    def reveal_type(x):
        print("reveal:", type(x))

class C: pass
class E(Exception): pass

def worker():
    print("Worker start")
    time.sleep(1)
    print("Worker end")
    raise C

x = ThreadPoolExecutor(4)
f = x.submit(worker)
try:
    r = f.result()
    print("R", r)
except BaseException as err:
    print("err =", err, type(err))
    e = f.exception()
    print("E", e.args)
```
And here's the traceback:
```
~/v27/bin/python x.py
Worker start
Worker end
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/threading.py", line 801, in __bootstrap_inner
    self.run()
  File "/usr/local/lib/python2.7/threading.py", line 754, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/Users/guido/v27/lib/python2.7/site-packages/concurrent/futures/thread.py", line 67, in _worker
    work_item.run()
  File "/Users/guido/v27/lib/python2.7/site-packages/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "x.py", line 18, in worker
    raise C
C: <__main__.C instance at 0x1025291b8>
```
At this point the program just hangs.

**Comment by rhansen** [comment]
> I've tried to write a simple demo program that raises an exception that doesn't derive from BaseException, and it doesn't even work -- the thread just crashes.

Which version of concurrent.futures are you using?  Support for old-style exceptions was only merged recently (agronholm/pythonfutures#58) and hasn't made it into a release yet.

**Comment by gvanrossum** [comment]
Oh, I used future 3.0.2.

Still I think that mypy has bigger problems than this when you are raising or catching exceptions that don't derive from `BaseError`, so I'm still pretty hesitant about supporting them here.

**Comment by rhansen** [comment]
Isn't mypy support for old-style exceptions somewhat orthogonal to the annotations in typeshed?  An alternative type checking tool might use the annotations from typeshed but behave differently than mypy with regard to old-style exceptions.

**Comment by gvanrossum** [comment]
Only "somewhat". What's your important use case?

**Comment by gvanrossum** [comment]
This PR seems stuck. @rhansen, do you still want to argue for this change? @ambv, what do you think?

**Comment by rhansen** [comment]
Yes, I would still like this to be merged—leaving it at BaseException doesn't match the behavior of the code.

Is there something I can do to reduce any negative impact this change might have on mypy?

**Comment by JelleZijlstra** [comment]
No, it only allows old-style classes and BaseException subclasses.

```
JZijlstra-MBA:cpython Jelle$ python2.7
Python 2.7.12 (default, Oct 11 2016, 14:42:58) 
[GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> class Foo: pass
... 
>>> class New(object): pass
... 
>>> raise Foo
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
__main__.Foo: <__main__.Foo instance at 0x109ca58c0>
>>> raise New
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: exceptions must be old-style classes or derived from BaseException, not type
```

For what it's worth, typing support for deprecated Python 2 features like old-style classes doesn't seem particularly high-priority to me either.

**Comment by rhansen** [comment]
I switched to Any and updated the commit message.

**Comment by gvanrossum** [comment]
OK, this sounds fine. I'll merge it once the tests have passed.

PS. In the future if you're responding to a code review, please push a new commit rather than amending the old one -- the discussion is easier to follow if updates are represented by their own commits, and we squash-merge everything anyway into master.

**Pull request merged**